engine/src/main/java/org/hibernate/validator/internal/engine/valueextraction/ValueExtractorManager.java

```diff
@@ -164,25 +223,58 @@ public ValueExtractorDescriptor getMaximallySpecificAndContainerElementCompliant
 			}
 		}
 
-		Set<ValueExtractorDescriptor> maximallySpecificContainerElementCompliantValueExtractors =
-				getMaximallySpecificValueExtractors( valueType, containerElementCompliantExtractors );
+		return containerElementCompliantExtractors;
+	}
 
-		if ( maximallySpecificContainerElementCompliantValueExtractors.isEmpty() ) {
-			return null;
-		}
-		else if ( maximallySpecificContainerElementCompliantValueExtractors.size() == 1 ) {
-			return maximallySpecificContainerElementCompliantValueExtractors.iterator().next();
-		}
-		else {
-			throw LOG.getUnableToGetMostSpecificValueExtractorDueToSeveralMaximallySpecificValueExtractorsDeclaredException( valueType,
-					ValueExtractorHelper.toValueExtractorClasses( maximallySpecificContainerElementCompliantValueExtractors ) );
+	/**
+	 * Returns the set of potentially type-compliant and container-element-compliant value extractors or an empty set if none was found.
+	 * <p>
+	 * A value extractor is potentially runtime type compliant if it might be compliant for any runtime type that matches the declared type.
+	 */
+	private Set<ValueExtractorDescriptor> getPotentiallyRuntimeTypeCompliantAndContainerElementCompliantValueExtractors(Type declaredType,
+			TypeVariable<?> typeParameter) {
+		boolean isInternal = TypeVariables.isInternal( typeParameter );
+		Type erasedDeclaredType = TypeHelper.getErasedReferenceType( declaredType );
+
+		Set<ValueExtractorDescriptor> typeCompatibleExtractors = valueExtractors.values()
+				.stream()
+				.filter( e -> TypeHelper.isAssignable( erasedDeclaredType, e.getContainerType() ) )
+				.collect( Collectors.toSet() );
```

Comment:
This is a transient set, it's not reused after that so I don't think it makes sense to make it immutable here.

We usually only make collections immutable when we store them in a field.